src/csharp/Grpc.Tools/ProtoCompile.cs

```diff
@@ -433,6 +439,18 @@ protected override string GenerateResponseFileCommands()
             return cmd.ToString();
         }
 
+        // If possible, disambiguate output dir by adding a hash of the proto file's path
+        static string MaybeEnhanceOutputDir(string outputDir, ITaskItem[] protobufs)
```

Comment:
Currently, the directory hash will always be added by Grpc.Tools. Two places where I'm not certain:
- should we add a way of disabling this new functionality? (currently it's always on and that could perhaps cause trouble for some users. E.g. the ones that are generating the files outside of the obj/ dir and rely on the path of the created files)
- For files with no relative dir (rooted in the project root), should we still include the hash? In this case we don't need to prefix the files and their output path would stay the same as before this PR - so for simple projects, this could save some output dir cleanup (that might otherwise be necessary when upgrading to the newer Grpc.Tools)

Reply:
I can add an option if you like. Any preferred naming for the option?

Is files in the project root really a thing? I thought there is always at least a protos folder.

Reply:
Let's add the option to make it possible to disable the hashed paths when needed (= which is basically fallback to the old behavior)
Ad naming: how about `Protobuf_AvoidGeneratedFileCollisions`, `Protobuf_DisambiguateGenerateFiles` or sth in that sense?
The options can be on by default (using hashes seems like a good default behavior) and one could disable it if needed.

Reply:
Is it OK to only have it on property level or do you also want to have it on the item level? I think property level should be enough.

Reply:
I just saw that in `CppGeneratorServices`, we already use `GetRelativeDir` to generate intermediate paths instead of hashes. `protoc` seems to generate the folders by default for C++.

Do we want to use hashes for C# or do we also want to just create the folders using the `GetRelativeDir` approach? Right now `MaybeEnhanceOutputDir` in `ProtoCompile` is always applied, also for C++. We would need to disable that based on `Generator`.

Technically, we could do it either way. But using the relative dir also in C# would make both generators more similar. And C++ would also already have the problem of the paths getting long. So having this behavior also in C# wouldn't create a gap.

Reply:
I found another bug in the Cpp generator. `Protobuf_PrepareClean` expects to find the expected outputs based on `Protobuf` instead of `Protobuf_Compile`. But as `ProtoRoot` is missing, the paths are not resolved correctly for out-of-project protos and cleanup won't find the files.